### PR TITLE
Implement the session tracker.

### PR DESCRIPTION
**Pull request by kinbiko**
> Note: The base branch of this PR will later be changed to `session-tracking`. Currently pointing to `session-tracking-delivery-logic`, as the code on this branch is necessary for this change and that branch has not yet been merged.

## Goal

Enable session tracking through `bugsnag.StartSession(ctx)`.

### Added

- `bugsnag.StartSession()`
- Session specific sub-config, mainly in order to encapsulate and build the sessions package.

### Changed

- Refactored session related code into its own package, and encapsulated.
- Moved the headers logic - which is shared between the standard report payload and the session payload - to its own package to avoid cyclic dependencies between `bugsnag` and `sessions`.
- Move session delivery functions to `publisher` methods.
- use `httptest` over manually creating a HTTP server.

## Tests

Fairly well tested. There's also a fairly comprehensive integration test for manual session tracking.

### Alternative Approaches

I found that if I kept doing sessions in the `bugsnag` package, the global scope got populated quickly and I ended up writing Java-esque functions and types. I therefore created the `sessions` subpackage.

### Outstanding Questions

I'm not super happy with having to expose the `DefaultSessionPublishInterval` but it's the best thing I could think of - that wouldn't require extensive refactoring - that would allow me to run the integration test in a reasonable amount of time. Feedback welcome.

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: sessions/tracker.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+func (s *sessionTracker) StartSession(ctx context.Context) context.Context {
+	session := newSession()
+	s.sessionChannel <- session
+	return context.WithValue(ctx, contextSessionKey, session)
+}
+
+func (s *sessionTracker) interval() time.Duration {
+	s.config.mutex.Lock()
```
Comment (CodeHex) [inline sessions/tracker.go]: Is there any other places where we need to lock the config? Looking at the session tracker and publisher structs, they seem to access the config often. Could this be modified underneath them by the user?

File: bugsnag.go [hunk 2]
```diff
@@ -152,3 +169,21 @@ func init() {
...
+		Version:         VERSION,
+		PublishInterval: DefaultSessionPublishInterval,
+		Transport:       Config.Transport,
+		ReleaseStage:    Config.ReleaseStage,
+		Hostname:        Config.Hostname,
+		AppType:         Config.AppType,
+		AppVersion:      Config.AppVersion,
+		Logger:          Config.Logger,
+	})
+	if sessionTracker == nil {
```
[the remark this answers is not in the document]
Comment (kinbiko) [reply bugsnag.go]: I think a warning message is fair at this point, and is a lot less breaking in terms of changes required. Let me know if you disagree.

File: sessions/tracker.go (hunk 1 above)
Comment (kinbiko) [reply sessions/tracker.go]: I've added a lock around publishing. I had a look around the sessions package and I think this is the only location where we need to worry about a race condition.
